Baip 254/multiple cars (#26)
bringauto/external-server: README.md

# External Server

This directory contains a external server that communicates with an external client, which is part of the [Module Gateway](https://gitlab.bringauto.com/bring-auto/fleet-protocol-v2/module-gateway).

> # Fleet protocol deviations

Reviewer: This can be removed

This implementation of the External Server handles multiple cars registered under a single company.

# Requirements

- Python (version >= 3.10)

# Usage

## Install dependencies

### Python packages

Install the required Python packages in virtual environment by running the following

```bash
python3 -m venv .venv && \
source .venv/bin/activate && \
pip3 install -r requirements.txt
```

### Submodules

Update the [fleet protocol](https://github.com/bringauto/fleet-protocol) submodule and compile the protobuf

```bash
git submodule update --init lib/fleet-protocol && \
pushd lib/fleet-protocol/protobuf && \

> find ./definition -name "*.proto" -exec protoc -I=./definition --python_out=./compiled/python --pyi_out=./compiled/python {} +

Reviewer: The files are already compiled in the repo. This is so the user doesn't have to install the correct protoc version every time. Or did you ran into any issues with that?

Author: You need to also include the .pyi files to allow for static type checking (but yes, it is not necessary for running the external server).

popd
```

## Configure the External Server

Prepare your config file for the External Server. The config file can be found in `config/config.json`.

> As an example of a filled-up config file, see the `config/config.json.example`. When using the modified `config/config.json.example`, remove the `example` suffix from the file name.

Reviewer: Isnt it better to use example.json and just giving the right path when starting the script?

Author: OK, the config.json.example is now replaced with config.json, that can be used directly

### Server configuration

Set up the logging, the MQTT connection parameters and company name and the External server behavior.

- `logging` - contains the following keys:
  - `log-path` - path to the directory where logs will be stored.
  - `verbosity` - if `True`, the server will print logs to the console and set the logging level to `DEBUG`. If `False`, the server will only print logs to the console and set the logging level to `INFO`.
- `company_name` - used for MQTT topics name, should be same as in module gateway; only lowercase characters, numbers and underscores are allowed.
- `mqtt_address` - IP address of the MQTT broker.
- `mqtt_port` - port of the MQTT broker.
- `mqtt_timeout` (in seconds) - timeout for getting a message from MQTT Client.

> - `timeout` (in seconds) - Maximum time amount between Status and Command messages.

Reviewer: In the external server, status and command are not request-response. Isn't this timeout between message and message response?

Author: Indeed, the description is misleading. fixed

- `send_invalid_command` - sends command to Module gateway even if External Server detects invalid command returned from external_server_api; affects only normal communication.
- `sleep_duration_after_connection_refused` - if the connection to Module Gateway was refused, the External Server will sleep for a defined duration before the next connection attempt proceeds.

### Common modules

> One of the last items in the config file is `common_modules`, represented by key-value pairs. The key is the ID of the module, the value contains following

Reviewer: What is ID? Module number?

Author: It is.

- `lib_path` (required) - path to module shared library (`*.so`).
- `config` (optional) - specification of config for the module, any key-value pairs will be forwarded to module implementation init function; when empty or missing, empty config forwarded to init function.

A common module will be used for all cars. No such module can be defined in the car configuration.

See the `config/config.json.example` for an example of modules configuration.

### Cars

> The last item in the config file is `cars`, represented by key-value pairs. The key is the name of the car, the value is a dictionary containing car-specific modules keyed as `specific_modules`.

Reviewer: Do I use cars only if i want to add some specific modules to specific cars?

Author: Car names have to be always defined. The dictionary next to car name can be left empty, if no car-specific modules have to be defined.

The structure of the `specific_modules` is the same as the `common_modules` structure.

See the `config/config.json.example` for an example of car configuration.

Configuring module with the same ID both in `common_modules` and `specific_modules` is invalid and the server will not start.

Note that for each car, at least one module has to be defined, either in `common_modules` or `specific_modules`.

## Start the External Server

After configuration and installation of the dependencies, run External Server with this command:

```bash
python3 external_server_main.py --config <str> [--tls] [--ca <str>] [--cert <str>] [--key <str>]
```

- `-c or --config <str>` = path to the config file, default = `./config/config.json`
- `--tls` = tls mqtt authentication

Following arguments are used if argument `tls` is set:

- `--ca <str>` = path to ca certification

> - `--cert <str>` = path to cert file"

Reviewer: remove trailing quote

- `--key <str>` = path to key file

# Unit tests

## Necessary steps before testing

First, do the steps from the [Install dependencies](#install-dependencies) section.

The proceed with the following steps.

### Install the external server package

Install the package in editable mode and install test requirements (assuming you already installed the requirements for the server):

```bash
pip install -e .
pip install -r tests/requirements.txt
```

Update submodules

```bash
git submodule update --init --recursive
```

### Install the shared library

Compile a shared library for the [Example Module](https://github.com/bringauto/example-module/). This requires

- the CMakelib installed (see [here](https://github.com/cmakelib/cmakelib)) and the `CMLIB_DIR` env variable set to the installation directory and exported,
- the [example-module](https://github.com/bringauto/example-module/) cloned as a submodule in the `tests/utils` directory.

Run the following

```bash
pushd tests/utils/example_module && \

> if [ ! -d "_build" ]; then mkdir _build; fi && \

Reviewer: `mkdir -p _build` is for ignoring already existing directories

Author: Used

cd _build && \

> cmake .. -DCMLIB_DIR=https://github.com/cmakelib/cmakelib && \

Reviewer: add Cmlib to test requirements

Author: Done (line 125)

# External Server

> This directory contains a external server that communicates with an external client, which is part of the [Module Gateway](https://gitlab.bringauto.com/bring-auto/fleet-protocol-v2/module-gateway).

Reviewer 2: This should point to github not internal gitlab

> This directory contains a external server that communicates with an external client, which is part of the [Module Gateway](https://gitlab.bringauto.com/bring-auto/fleet-protocol-v2/module-gateway).

Reviewer 2: Don't start readme with "This directory contains" 😀

The External Server handles multiple cars registered under a single company.

# Requirements

- Python (version >= 3.10)

# Usage

## Install dependencies

### Python packages

Install the required Python packages in virtual environment by running the following

```bash
python3 -m venv .venv && \
source .venv/bin/activate && \
pip3 install -r requirements.txt
```

### Submodules

Update the [fleet protocol](https://github.com/bringauto/fleet-protocol) submodule

```bash
git submodule update --init lib/fleet-protocol
```

## Configure the External Server

Prepare your config file for the External Server. The config file can be found in `config/config.json`.

As an example of a filled-up config file, see the `config/config.json`. Before running the server, update the `config/config.json` accordingly.

### Server configuration

Set up the logging, the MQTT connection parameters and company name and the External server behavior.

- `logging` - contains the following keys:
  - `log-path` - path to the directory where logs will be stored.

>   - `verbosity` - if `True`, the server will print logs to the console and set the logging level to `DEBUG`. If `False`, the server will only print logs to the console and set the logging level to `INFO`.

Reviewer 2: Little lost what will be printed when to be honest.

Author: Fixed

- `company_name` - used for MQTT topics name, should be same as in module gateway; only lowercase characters, numbers and underscores are allowed.
- `mqtt_address` - IP address of the MQTT broker.
- `mqtt_port` - port of the MQTT broker.
- `mqtt_timeout` (in seconds) - timeout for getting a message from MQTT Client.
- `timeout` (in seconds) - Maximum time amount between Status or Command messages and receiving corresponding responses.
- `send_invalid_command` - sends command to Module gateway even if External Server detects invalid command returned from external_server_api; affects only normal communication.
- `sleep_duration_after_connection_refused` - if the connection to Module Gateway was refused, the External Server will sleep for a defined duration before the next connection attempt proceeds.

### Common modules

> One of the last items in the config file is `common_modules`, represented by key-value pairs. The key is the module ID (a module number), the value contains following

Reviewer 2: ```suggestion
One of the last items in the config file is `common_modules`, represented by key-value pairs. The key is the module ID (a module number), the value contains following:
```

- `lib_path` (required) - path to module shared library (`*.so`).

> - `config` (optional) - specification of config for the module, any key-value pairs will be forwarded to module implementation init function; when empty or missing, empty config forwarded to init function.

Reviewer 2: ```suggestion
- `config` (optional) - module specific configuration, any key-value pairs will be forwarded to module implementation init function. When empty or not provided, empty configuration is forwarded to init function of module.
> For configuration of respective module see documentation of that module.
```

> A common module will be used for all cars. No such module can be defined in the car configuration.

Reviewer 2: ```suggestion
> [!WARNING]
> A common module will be used for all cars. No such module can be defined in the car configuration.
```

See the `config/config.json` for an example of modules configuration.

### Cars

The last item in the config file is `cars`, represented by key-value pairs. The key is the name of the car, the value is a dictionary containing car-specific modules keyed as `specific_modules`.

The structure of the `specific_modules` is the same as the `common_modules` structure.

See the `config/config.json` for an example of car configuration.

> Configuring a module with the same ID both in `common_modules` and `specific_modules` is invalid and the server will not start.

Reviewer 2: ```suggestion
> [!WARNING]
> Configuring a module with the same ID both in `common_modules` and `specific_modules` is invalid and the server will not start.
```

Note that for each car, at least one module has to be defined, either in `common_modules` or `specific_modules`.

> 

Reviewer 2: ```suggestion
> [!IMPORTANT]
> Keep in mind that for each car, at least one module has to be defined, either in `common_modules` or `specific_modules`.

```

# External Server

This directory contains a external server that communicates with an external client, which is part of the [Module Gateway](https://gitlab.bringauto.com/bring-auto/fleet-protocol-v2/module-gateway).

> The External Server handles multiple cars registered under a single company.

Author: Fixed
